src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcTransport.cs

```diff
@@ -0,0 +1,270 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Pipes;
+using System.Linq;
+using System.Net;
+using System.Net.Sockets;
+using System.Runtime.InteropServices;
+using System.Security.Principal;
+using System.Threading;
+
+namespace Microsoft.Diagnostics.NETCore.Client
+{
+    /// <summary>
+    /// An endpiont used to acquire the diagnostics stream for a runtime instance.
+    /// </summary>
+    internal interface IIpcEndpoint
+    {
+        /// <summary>
+        /// Checks that the client is able to communicate with target process over diagnostic transport.
+        /// </summary>
+        /// <returns>
+        /// True if client is able to communicate with target process; otherwise, false.
+        /// </returns>
+        bool CheckTransport();
```

Comment:
I'll update the comments on the method to better reflect what it's trying to do. I'm thinking of refactoring the method to `Task WaitForConnectionAsync(CancellationToken)` so that it will return a completed task only when a viable connection becomes available. Consumers of IIpcEndpoint would still need to call Connect in order to get the stream.

I'm wondering if we need to morph the API a bit more to effectively force everything that wants the stream to call something like `Task<Stream> ConnectionAsync(CancellationToken)` so that way we can avoid having multiple threads calling WaitForConnectionAsync and races of 'who calls Connect first'; if all of the threads see that WaitForConnectionAsync completes, and one of them consumes the stream, then others will get blocked by the reset event waiting for another stream.

Reply:
Regarding PidIpcEndpoint, I'm not sure what we should do if the socket file is abandoned for whatever reason.

Presumably, it is already the newest socket file for that process, since get the files that match the pid and then take the "latest" one, so there is no other socket file to try for that pid.

I could change PidIpcEndpoint to open the stream to the socket, cache it internally, and then hand out the stream from the Connect method, just like the ServerIpcEndpoint does (if I do that then maybe these classes converge and really only differ on how they receive the stream, either from the reversed server or getting it itself). It would then only open a new stream on demand (demanded by WaitForConnectionAsync or Connect methods) after handing out the previous stream. Then I can apply the same check behavior that ServerIpcEndpoint.WaitForConnectionAsync provides so that it should indefinitely wait on that connection; possibly do some specialization such that it can check if there is a newer socket file, and if there is, abandon the existing one and start monitoring the new one.